docs: add GPU module guidelines (#4142)
nf-core/website: sites/docs/src/content/docs/specifications/components/modules/software-requirements.md

      error("This module does not support Conda. Please use Docker / Singularity / Podman instead.")
  }
  ```

### Dual-container pattern

When the GPU container is substantially larger, modules SHOULD switch between containers based on `task.accelerator`:

```groovy
conda "${ task.accelerator ? "${moduleDir}/environment.gpu.yml" : "${moduleDir}/environment.yml" }"
container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
    (task.accelerator ? '<singularity-gpu-url>' : '<singularity-cpu-url>') :
    (task.accelerator ? '<docker-gpu-url>' : '<docker-cpu-url>') }"
```

A separate `environment.gpu.yml` SHOULD be provided for GPU-specific dependencies.
The CPU `environment.yml` MUST remain unchanged so that non-GPU users are unaffected.

GPU containers SHOULD be built using [Wave](https://wave.seqera.io) from the `environment.gpu.yml` file.
Both Docker and Singularity URLs MUST be provided.

:::note{title="Wave build template"}
Pass `--build-template conda/micromamba:v2` to Wave when building GPU environments. Required for now until it becomes the default.
:::

> ### CUDA version pinning

Reviewer: This section is getting extremely long/wordy, more about highly specific implementations that I don't feel fits in the specifications and is almost turning into a tutorial (do you agree @christopher-hakkaart ?)  

At a minimum I feel you could move most of this to a standalone 'GPU' section of the docs in a separate page, but also I think you should maybe move much of this to a reference guide under the 'development' section of the website somewhere.

The only actual requirmenet here is 'Pin and report cuda-versions' in you your CUDA related GPU modules. Everything else is not a specfiication.

Reviewer 2: Agree with @jfy133 here. It's reasonably verbose, and I would like to keep the specifications as specifications.

I've made a PR to split them here: https://github.com/nf-core/website/pull/4251

- If the multi-tool container already exists, use [this](https://midnighter.github.io/mulled) helper tool to obtain the `mulled-*` path.

## GPU-capable modules

The container approach depends on the tool:

> - **Significant GPU overhead** (e.g., CUDA PyTorch adds ~3 GB): use the dual-container pattern below. For example, [`ribodetector`](https://github.com/nf-core/modules/tree/master/modules/nf-core/ribodetector).

Reviewer: I don't really follow what these are referring to, I think you're missing some introducing context here of what we the specs are trying solve for this section of a module, e.g, GPU containers can be very large and sometimes can be proprietary, therefore we specify different ways to account for these factors.

For example, by dual container approaches are you talking about where there is a CPU/GPU variant of a tool?

Author: Yeah, so Ribodetctor has a GPU module, but you have to install all the GPU gubbins alongside, which is a waste for folks who don't want / need it.

Author: Good point, added an intro paragraph in b2087ec explaining the two awkward properties (GPU builds can be several GB larger, some vendor containers are proprietary with no conda equivalent) and why that motivates three container approaches. Also reworded the first bullet to make the CPU/GPU dual-variant case explicit rather than burying it behind "significant GPU overhead".

  ```groovy
  if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
      error("This module does not support Conda. Please use Docker / Singularity / Podman instead.")
  }
  ```

### Dual-container pattern

When the GPU container is substantially larger, modules SHOULD switch between containers based on `task.accelerator`:

```groovy
conda "${ task.accelerator ? "${moduleDir}/environment.gpu.yml" : "${moduleDir}/environment.yml" }"
container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
    (task.accelerator ? '<singularity-gpu-url>' : '<singularity-cpu-url>') :
    (task.accelerator ? '<docker-gpu-url>' : '<docker-cpu-url>') }"
```

A separate `environment.gpu.yml` SHOULD be provided for GPU-specific dependencies. The CPU `environment.yml` MUST remain unchanged so that non-GPU users are unaffected.

GPU containers SHOULD be built using [Wave](https://wave.seqera.io) from the `environment.gpu.yml` file. Both Docker and Singularity URLs MUST be provided.

### CUDA version targeting

> NVIDIA drivers are backward compatible: a host with a CUDA 12.x driver can run containers built for CUDA 11.x or 12.x. A CUDA 11.8 container works on any modern GPU host. The reverse is not true: a CUDA 12.x container cannot run on a host with only a CUDA 11.x driver.

Reviewer: ```suggestion
NVIDIA drivers are backward compatible: a host with a CUDA 12.x driver can run containers built for CUDA 11.x or 12.x. A CUDA 11.8 container works on any modern GPU host.
The reverse is not true: a CUDA 12.x container cannot run on a host with only a CUDA 11.x driver.
```

Author: Addressed in b2087ec alongside your "too much detail" comment below — the section is now much shorter ("CUDA version pinning"), and the driver-compatibility note is on two lines as suggested.

Within a major version there is full forward compatibility: a container built with any CUDA 12.x works on any CUDA 12.0+ driver.

> In practice, **CUDA 12.x is the best default** for new modules: it gives access to the latest tool versions (newer PyTorch and similar packages are dropping CUDA 11 builds) while remaining compatible with all current GPU cloud instances and recent HPC drivers. A CUDA 11.8 alternative can be provided for users on older systems.

Reviewer: This section is going into too much detail for specifications

Author: Agreed, trimmed in b2087ec. Kept only the CUDA major-version pinning guidance (which is the actionable spec point) plus a two-line note on driver compatibility. Dropped the PyTorch/cloud-instance commentary and the `meta.yml` aside.

- If the multi-tool container already exists, use [this](https://midnighter.github.io/mulled) helper tool to obtain the `mulled-*` path.

## GPU-capable modules

GPU-enabled software has two properties that make it awkward to package the same way as CPU-only tools: GPU builds (e.g., CUDA PyTorch) can be several GB larger than their CPU counterparts, and some vendor containers (e.g., NVIDIA Parabricks) are proprietary with no conda equivalent.
The specification therefore allows three container approaches, chosen according to the tool:

- **Dual CPU/GPU variants of a tool**, where the GPU build has significant overhead (e.g., CUDA PyTorch adds ~3 GB): use the [dual-container pattern](#dual-container-pattern) below so CPU-only users are not penalised.
  For example, [`ribodetector`](https://github.com/nf-core/modules/tree/master/modules/nf-core/ribodetector).
- **Minimal GPU overhead or CPU fallback within one container**: a single container is simpler and preferred.
- **Vendor-provided GPU containers** with no conda equivalent: use the vendor container directly.

>   For example, [`parabricks/rnafq2bam`](https://github.com/nf-core/modules/tree/master/modules/nf-core/parabricks/rnafq2bam) uses NVIDIA's container.

Reviewer: ```suggestion

  For example, [`parabricks/rnafq2bam`](https://github.com/nf-core/modules/tree/master/modules/nf-core/parabricks/rnafq2bam) uses NVIDIA's container.
```

Author: Applied in 951a232.

>   These modules SHOULD guard against conda/mamba profiles:

Reviewer: ```suggestion
  Vender-provided container modules SHOULD guard against conda/mamba profiles:
```

Author: Applied in 951a232 (using "Vendor-provided container modules" — assumed "Vender" was a typo).

```groovy
conda "${ task.accelerator ? "${moduleDir}/environment.gpu.yml" : "${moduleDir}/environment.yml" }"
container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
    (task.accelerator ? '<singularity-gpu-url>' : '<singularity-cpu-url>') :
    (task.accelerator ? '<docker-gpu-url>' : '<docker-cpu-url>') }"
```

A separate `environment.gpu.yml` SHOULD be provided for GPU-specific dependencies.
The CPU `environment.yml` MUST remain unchanged so that non-GPU users are unaffected.

GPU containers SHOULD be built using [Wave](https://wave.seqera.io) from the `environment.gpu.yml` file.
Both Docker and Singularity URLs MUST be provided.

### CUDA version pinning

The `environment.gpu.yml` SHOULD pin the CUDA major version to avoid the conda solver selecting builds for unreleased CUDA versions:

```yaml
dependencies:
  - "bioconda::ribodetector=0.3.3"
  - "conda-forge::pytorch-gpu=2.10.0"
  - "conda-forge::cuda-version>=12,<13"
```

NVIDIA drivers are backward compatible, so a container built against CUDA 12.x will run on any CUDA 12.0+ host driver.

> A CUDA 12.x container will not run on an older CUDA 11.x driver, so modules that need to support older hosts MAY provide an alternative `environment.gpu.yml` pinned to CUDA 11.

Reviewer: But you said they are backward compatible 🤔

Author: Fair point — the original phrasing read like it contradicted itself. Reworded in 951a232 to: "NVIDIA drivers are backward compatible with older CUDA versions: a host with a CUDA 12.x driver can run containers built against CUDA 11.x or 12.x. The reverse is not supported, so a CUDA 12.x container cannot run on a host with only a CUDA 11.x driver." NVIDIA's actual term "backward compatible" here means "newer driver runs older containers" only; making that direction explicit should clear up the confusion.

> ### Script patterns

Reviewer: Can you provide a more specific title?

Author: Renamed to "Binary and GPU count selection in scripts" in 951a232.

> Tools that provide separate GPU and CPU binaries SHOULD select between them based on `task.accelerator`. For example, [`ribodetector`](https://github.com/nf-core/modules/blob/master/modules/nf-core/ribodetector/main.nf):

Reviewer: ```suggestion
Tools that provide separate GPU and CPU binaries SHOULD select between them based on `task.accelerator`.
For example, [`ribodetector`](https://github.com/nf-core/modules/blob/master/modules/nf-core/ribodetector/main.nf):
```

Reviewer: And this example URL should use a hash commit in this case

Author: Applied in 951a232 — split onto two lines and pinned the ribodetector URL to `20423f5`. Also pinned the parabricks example URL below to `0c44f69` for consistency.

```groovy
def binary = task.accelerator ? "ribodetector" : "ribodetector_cpu"
```

> Tools that accept a GPU count SHOULD read it from `task.accelerator.request`, allowing users to override via their pipeline config (e.g., `accelerator = 2`). For example, [`parabricks/rnafq2bam`](https://github.com/nf-core/modules/blob/master/modules/nf-core/parabricks/rnafq2bam/main.nf):

Reviewer: ```suggestion
Tools that accept a GPU count SHOULD specify this in the command using `task.accelerator.request`, allowing users to override via their pipeline config (e.g., `accelerator = 2`).
This parameter SHOULD NOT be hardcoded.

For example, [`parabricks/rnafq2bam`](https://github.com/nf-core/modules/blob/master/modules/nf-core/parabricks/rnafq2bam/main.nf):
```

Author: Applied in 951a232.
